src/traces/sankey/defaults.js

```diff
@@ -31,13 +31,13 @@ module.exports = function supplyDefaults(traceIn, traceOut, defaultColor, layout
         return Color.addOpacity(defaultNodePalette(i), 0.8);
     }));
 
-    coerce('link.label');
     coerce('link.source');
     coerce('link.target');
     coerce('link.value');
     coerce('link.line.color');
     coerce('link.line.width');
 
+    coerce('link.label', traceOut.link.value.map(function() {return '';}));
```

Comment:
🐎  (also for `link.color`) we're always doing this `map` even if we don't need the default. Would it be worth either:
- running that after the `coerce` call, ie:
```js
var linkLabel = coerce('link.label');
if(!linkLabel) traceOut.link.label = traceOut.link.value.map(function() {return '';});
```
- building into `coerce` that you can pass a function as `dflt` and it will only be called if needed?
```js
coerce('link.label', function() { return traceOut.link.value.map(function() {return '';}); });
```
thoughts?

Reply:
Also typically when we rely on a previously coerced value to modify later behavior, usually we just save it from its coercion: `var linkValue = coerce('link.value')`

Which brings up the question: some of these inputs are strictly necessary in order to have meaningful output. What's the minimal set here, without which we should [automatically hide the trace](https://github.com/plotly/plotly.js/blob/790d470b901a38713728b5c9dc10bc624fed39ad/src/traces/scatter/defaults.js#L34-L37)? If there's no `link.value` this `map` will throw an error, which we definitely don't want.

Reply:
Oh I guess you default to `[]` so it won't error... but it still doesn't seem meaningful right?

Reply:
Good point, definitely worth avoiding useless work. Whenever adding some array attributes, I felt it'd be good to be able to specify a default value for the array elements (in `attributes.js`) rather than just the whole array. But your suggestion is more flexible because it allows a function (default element values depending on other values) so I'll

Reply:
@monfera why don't you move this piece of logic to the `calc` step?

Reply:
@etpinard I think of array values as conceptually default-able (though there's no `elementDefault` in `attributes.js`), and saw usages elsewhere that were also in `defaults.js`. But I can do it in `calc` or even in the tooltip caller code `linkLabel || ''` - after all, only the tooltip has a problem with `undefined`, which is external to `sankey`. As @alexcjohnson had an alternative suggestion, I wonder which direction would be best in this case.

Reply:
Ah, if you can handle it entirely at the point of hover - so no fallback processing needs to happen at all during plotting - that seems like the best of all, performance-wise, and we do have other examples of arrays that may or may not exist and we leave it until the value is needed to test this rather than trying to fill it in at `supplyDefaults` or even `calc`.

Reply:
The fix got pivoted to this consensus, the `assertLabel` test function now also checks whether the value no longer breaches the tooltip bounding box for a new test case (and other tests that use it already). The test properly breaks when I comment out the fix.

Reply:
The commit is in, that's all there is to it, plus the tests 😄